signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,592 @@
+
+        public SignalFlowExceptionStreamMessage(final SignalFlowException exception) {
+            super("error", null, exception.getMessage());
+            this.exception = exception;
+        }
+
+        public SignalFlowException getException() {
+            return this.exception;
+        }
+    }
+
+    /**
+     * Value Object that handles binary data message conversion in construction
+     */
+    protected static class TransportDataMessage {
+
+        protected byte version;
+        protected byte messageType;
+        protected String channelName;
+        protected long logicalTimestampMs;
+        protected ArrayList<Map<String, Object>> data = new ArrayList<Map<String, Object>>();
+
+        public TransportDataMessage(byte[] data) {
+            try {
+                this.version = data[0];
+                this.messageType = data[1];
+                this.channelName = new String(data, 4, 16, "UTF-8");
+
+                if (messageType == 5) {
+                    byte[] logicalTimestampMsBytes = Arrays.copyOfRange(data, 20, 28);
+                    this.logicalTimestampMs = ByteBuffer.wrap(logicalTimestampMsBytes).getLong();
+
+                    byte[] payload = Arrays.copyOfRange(data, 32, data.length);
+                    for (int element = 0; element < (payload.length / 17); element++) {
+
+                        int index = element * 17;
+
+                        byte[] tsIdBytes = Arrays.copyOfRange(payload, index + 1, index + 9);
+                        String encodedTsId = StringUtils
+                                .remove(DatatypeConverter.printBase64Binary(tsIdBytes), "=");
+
+                        Map<String, Object> elementMap = new HashMap<String, Object>(4);
+                        elementMap.put("tsId", encodedTsId);
+
+                        byte[] valueBytes = Arrays.copyOfRange(payload, index + 9, index + 17);
+
+                        switch (payload[index]) {
+                        case 1: // long value
+                            elementMap.put("value", ByteBuffer.wrap(valueBytes).getLong());
+                            break;
+
+                        case 2: // double value
+                            elementMap.put("value", ByteBuffer.wrap(valueBytes).getDouble());
+                            break;
+
+                        default: // not suppose to happen
+                            log.warn("ignoring data message with unknown value type {}",
+                                    payload[index]);
+                            continue;
+                        }
+
+                        this.data.add(elementMap);
+                    }
+                } else {
+                    log.warn("Unsupported binary message type {}", messageType);
+                }
+            } catch (Exception ex) {
+                log.error("failed to construct transport data message", ex);
+            }
+        }
+
+        public int getVersion() {
+            return (int) version;
```

Comment:
If you have to cast here then you probably should have casted when reading the value from the data array.